### ISSUE TITLE
Qwen3-ForcedAligner-0.6B (qwen-asr) emits zero-duration word spans [t, t] on source audio at high rates

### ISSUE THREAD

**Issue by hasaki321** [QwenLM/Qwen3-ASR#197]
### Description

When running `Qwen3ForcedAligner` (`qwen-asr==0.0.6`, model `Qwen3-ForcedAligner-0.6B`) on **source** speech + transcript pairs, a large fraction of utterances contain at least one **lexical** alignment item with `start_time == end_time` (zero-duration span `[t, t]`).

**Related:** This overlaps with [#55 返回字时间戳有问题](https://github.com/QwenLM/Qwen3-ASR/issues/55) (closed as `not_planned` / inactive, not fixed). That report described Chinese character timestamps being all zeros, or several consecutive characters sharing the same `start_time`. We still observe the same class of failure on current `qwen-asr==0.0.6` / `Qwen3-ForcedAligner-0.6B`, and provide rates + a minimal reproduction below.

These are not punctuation-only tokens: the aligner strips punctuation before alignment, and the zero-duration items are real words / Chinese characters (e.g. English `to` / `a`, Chinese `我` / `的`). On waveform inspection, the corresponding speech energy is clearly non-zero (often ~100–300 ms), so `[t, t]` is not a faithful word boundary.

We evaluated **sampled subsets** of several speech corpora used in our ASR pipeline (Common Voice / CoVoST-style, Emilia-style, GigaSpeech-style, WenetSpeech-style), source language English or Chinese. For each clip we store `src_alignment_words = [{text, start_time, end_time}, ...]` from a successful `align()` call.

**Definitions (source side only):**

- **Sample rate:** a clip is counted if **any** FA item has `end_time == start_time`.
- **Word rate:** `#zero-duration items` / `#all FA items on src`.

**Observed rates:**

| Subset (corpus / direction) | #clips (approx.) | Sample rate | Word rate |
|----------------------------|------------------|------------:|----------:|
| Common Voice / CoVoST-style, en→zh (src=en) | ~205k | **22.87%** | **2.61%** |
| Common Voice / CoVoST-style, zh→en (src=zh) | ~3.6k | **7.31%** | **0.42%** |
| Emilia, en) | ~240k | **50.54%** | **3.39%** |
| Emilia, zh | ~226k | **29.64%** | **1.36%** |
| GigaSpeech, en | ~241k | **39.40%** | **3.06%** |
| WenetSpeech, zh | ~253k | **30.94%** | **2.81%** |

Aggregated: overall source sample rate **~35.0%**, word rate **~2.48%**; by language sample rate **src-en ~38.4%**, **src-zh ~30.2%**.

**Concrete badcase (source Chinese):**

- Transcript: `对于更高阶的导数，我们可以继续同样的过程。`
- ForcedAligner output includes `我 = [2.72, 2.72]` (0 ms), while neighbor `们 = [2.72, 2.88]`.
- Waveform around 2.72 s has clear energy; re-running FA still yields zero-duration items.
- MFA on the same clip recovers a positive span (e.g. `我们 ≈ [2.63, 2.92]`, ~290 ms). Across a Covost zh-en zero-boundary subset we checked, MFA matched all Qwen zero-duration lexical items with **strictly positive** duration.

<img width="1400" height="392" alt="Image" src="https://github.com/user-attachments/assets/242a2714-5513-4e9e-9041-7152ebbf18b0" />

**Expected vs actual:**

| | Expected | Actual |
|---|----------|--------|
| Lexical word / character span | `end_time > start_time` with a plausible acoustic duration | Frequently `end_time == start_time` |
| Reproducibility | Occasional noise | Reappears on re-alignment of the same clips |
| Comparison | Continuous aligners (e.g. MFA) give positive durations | MFA matched items are all strictly positive in our checks |

**Questions:**

1. Is zero-duration lexical output considered expected under the 80 ms hard-argmax + LIS postprocess?
2. Is there a recommended decoding / post-processing path (soft expectation over bins, min duration, clamp, etc.) to avoid `[t, t]` for real words?
3. If this is a known limitation, could the docs call out that `start_time == end_time` may occur for lexical units and how users should handle it?

Happy to provide a small public pack (wav + transcript + printed FA items) if helpful. This report is about the open-weight ForcedAligner via `qwen-asr` only (source-side zero-duration; timestamp-overshoot past audio length is out of scope here).

### Reproduction

```python
import torch
from qwen_asr import Qwen3ForcedAligner

aligner = Qwen3ForcedAligner.from_pretrained(
    "Qwen/Qwen3-ForcedAligner-0.6B",  # or local checkpoint
    dtype=torch.bfloat16,
    device_map="cuda:0",
)

# Replace with a local mono wav + matching transcript.
# Example transcript that reproduces [t,t] on our side:
#   "对于更高阶的导数，我们可以继续同样的过程。"
results = aligner.align(
    audio=["/path/to/source.wav"],  # or (np.ndarray, 16000)
    text=["对于更高阶的导数，我们可以继续同样的过程。"],
    language=["Chinese"],
)

for item in results[0]:
    print(f"{item.text}\t{item.start_time}\t{item.end_time}")
    if item.start_time == item.end_time:
        print("  ^^ zero-duration lexical span")
```

Batch setting used in our pipeline (optional, same API): multi-GPU worker processes with `CUDA_VISIBLE_DEVICES` per process, `batch_size=32`, DataLoader `num_workers=4`, audio converted to mono float32 16 kHz before `align()`, language set explicitly to `English` or `Chinese`.

Implementation notes relevant to the symptom (from `qwen_asr` ForcedAligner):

1. Text is tokenized; punctuation is stripped (`clean_token` keeps letters/numbers/`'` only).
2. Each unit gets two `<timestamp>` slots (start/end).
3. Timestamps use hard argmax over discrete bins (`timestamp_segment_time=80` ms), then LIS monotonicity repair.
4. Predicted times are not clamped to audio duration.

### Logs

```shell
# Badcase excerpt (source Chinese, Qwen3-ForcedAligner-0.6B)
# transcript: 对于更高阶的导数，我们可以继续同样的过程。

对      0.8     0.96
于      0.96    1.04
更      1.04    1.28
高      1.28    1.44
阶      1.44    1.68
的      1.68    1.76
导      1.76    2.08
数      2.08    2.4
我      2.72    2.72      <-- zero duration
们      2.72    2.88
可      2.88    3.12
以      3.12    3.44
继      3.44    3.6
续      3.6     3.84
同      3.84    4.08
样      4.16    4.24
的      4.24    4.32
过      4.32    4.56
程      4.56    4.72

# MFA on the same region (reference, not Qwen output):
# 我们 ≈ [2.63, 2.92]  (positive duration ~0.29s)

# English-style pattern also seen, e.g.:
# to      4.4     4.4
# a       0.0     0.0
```

### Environment Information

- **OS:** Ubuntu 22.04.3 LTS (Linux 5.15.x, x86_64)
- **Python:** 3.12.12
- **Package:** `qwen-asr==0.0.6`
- **Model:** `Qwen3-ForcedAligner-0.6B`
- **PyTorch:** 2.6.0+cu124
- **Transformers:** 4.57.6
- **GPU:** NVIDIA A800
- **CUDA:** 12.4 (PyTorch build `cu124`)
- **Inference backend:** HuggingFace Transformers via `Qwen3ForcedAligner.from_pretrained` + `align()` (not vLLM)
- **Runtime dtype / device:** `torch.bfloat16`, `device_map="cuda:0"`
- **Cluster setup (approx.):** multi-GPU workers (`CUDA_VISIBLE_DEVICES` per process), batch size 32, DataLoader `num_workers=4`, audio normalized to mono 16 kHz float32 before alignment

### Known Issue

- [x] The issue hasn't been already addressed in Documentation, Issues, and Discussions.

**Comment by gagabobo** [comment]
这个我也遇到了，中文的很少遇到，韩语和日语这种bad case很多，有解决办法吗？

**Comment by hasaki321** [comment]
它论文里面说的是用MFA打的伪标签，实际上MFA通常对字级别的分词支持很差，如果你用MFA分字的画也会有同样的问题，那qwen force align也会有这个问题。训练层面的问题还真没啥好办法解决




何荣燊
***@***.***

**Comment by Adi2K** [comment]
Confirming this on a different workload, with data that may narrow the mechanism.

We run the GGUF conversion of `Qwen3-ForcedAligner-0.6B` on **conversational** medical audio (two- and four-speaker consultations and meetings), in English and German. Our rates are substantially higher than the read-speech corpora above.

**Word rate, by speech type** (words with `end_time == start_time`, over all aligned words):

| clip | language | style | words | zero-duration |
|---|---|---|---|---|
| single speaker reading a speech | en | read | 22 | **0%** |
| read prompts | en | read | ~1.3k | **1.2%** |
| clinical dialogue | de | read/scripted | 259 / 2,344 | **0.77% / 1.83%** |
| general conversation | de | spontaneous | 785 | **3.18%** |
| doctor–patient consultation | en | spontaneous | 1,291 | **7.82%** |
| four-speaker meeting | en | spontaneous | 2,348 | **8.18%** |

So roughly **3× the aggregate word rate reported above** on spontaneous conversational English. The rate tracks speech rate monotonically — read speech at the bottom, fast overlapping conversation at the top — which is what you would expect if the cause is short words falling inside a single timestamp bucket rather than a failure to locate them.

**Three observations that may help isolate it:**

1. **It is always `end == start`, never `end < start`.** Across 327 instances we inspected individually, every one is an exact collision, not a corrupted or inverted span.

2. **The affected words are overwhelmingly short function words.** On the English consultation: `to` (14), `you` (7), `a` (7), `and` (6), `of` (5), `I` (4), `the` (4). The most common affected word length is two characters — matching the `to` / `a` examples in the original report.

3. **Quantisation is not a factor.** We measured 8-bit and 4-bit conversions of the same model on identical audio and they agree closely (consultation 6.9% vs 8.7%, meeting 7.5% vs 8.8%) — the behaviour is present in both, so it is not introduced by our conversion.

**On the mechanism.** The shipped model metadata declares `classify_num = 5000` and `timestamp_segment_time_ms = 80`, and the decode selects the highest-scoring bucket per boundary and multiplies by 80. A word whose speech begins and ends within the same 80 ms window therefore receives the same bucket index twice, and `end == start` follows by construction. That is consistent with the training labels being floor(start/80) and floor(end/80).

If that is the whole story, then `[t, t]` means "shorter than one 80 ms bucket" and is a legitimate, if lossy, answer. But the report above shows `[t, t]` emitted over 100–300 ms of clear speech energy, with MFA recovering a positive span — which is a different failure. **As shipped these two cases are indistinguishable to a consumer**, because the per-bucket scores are discarded once the argmax is taken.

**What would help downstream users most**, in rough order of value:

1. **Document that `end_time == start_time` is a possible output** and what it signifies. Several consumers, including ours, assume a positive duration.
2. **Expose a per-word confidence or the boundary score**, so "short word" can be told apart from "could not place it". Other aligners in this space return a per-word score; this model returns only text, start and end.
3. Optionally, a `minimum_duration` parameter of the kind NVIDIA's aligner ships, widening a collapsed span from the midpoint — with the documented caveat that it can make adjacent spans overlap.

Happy to share per-clip distributions or re-run against a specific configuration if that is useful.
